### PR TITLE
feat(lsp): add textDocument/rangesFormatting support to vim.lsp.buf.format()

### LINKED ISSUE #27293: [lsp] Add support for textDocument/rangesFormatting
### Problem

I'm not sure what the policy is about implementing unreleased LSP features, but LSP 3.18 contains `textDocument/rangesFormatting`, which formats multiple ranges with a single request. I think this would be a very nice addition. But I can understand if we first want to wait for the 3.18 spec to be released. Then we can reopen this issue when it's time :)

### Expected behavior

The `range` parameter of `vim.lsp.buf.format` should support taking a list of ranges (which would then be forwarded to rangesFormatting instead of rangeFormatting)

### PR DESCRIPTION
**Pull request by tom-anders**
While this relies on a proposed LSP 3.18 feature, it's fully backwards compatible, so IMO there's no harm in adding this already.

Looks like some servers already support for this (e.g. gopls: https://go-review.googlesource.com/c/tools/+/510235), and for clangd there's an open PR: https://github.com/llvm/llvm-project/pull/80180

Fixes #27293

### CODE REVIEW

File: runtime/doc/lsp.txt [hunk 1]
```diff
@@ -1256,9 +1256,11 @@ format({options})                                       *vim.lsp.buf.format()*
                      name (client.name) matching this field.
                    • range (table|nil) Range to format. Table must contain
                      `start` and `end` keys with {row,col} tuples using (1,0)
-                     indexing. Defaults to current selection in visual mode
-                     Defaults to `nil` in other modes, formatting the full
-                     buffer
+                     indexing. Can also be a list of tables that contain
+                     `start` and `end` keys as described above, in which case
+                     textDocument/rangesFormatting support is required.
```
Comment (MariaSolOs) [inline runtime/doc/lsp.txt:1261]:
```suggestion
                     `textDocument/rangesFormatting` support is required.
```

File: runtime/doc/lsp.txt [hunk 2]
```diff
@@ -1256,9 +1256,11 @@ format({options})                                       *vim.lsp.buf.format()*
...
+                     Defaults to current selection in visual mode Defaults to
```
Comment (MariaSolOs) [inline runtime/doc/lsp.txt:1262]:
```suggestion
                     Defaults to current selection in visual mode and to
```

File: runtime/doc/news.txt [hunk 3]
```diff
@@ -206,6 +206,9 @@ The following new APIs and features were added.
     the original LSP `Location` or `LocationLink`.
   • Added support for connecting to servers using named pipes (Windows) or
     unix domain sockets (Unix) via |vim.lsp.rpc.domain_socket_connect()|.
+  • |vim.lsp.buf.format()| now supports formatting passing a list of ranges
+    via the `range` parameter (this requires support for the proposed
```
Comment (MariaSolOs) [inline runtime/doc/news.txt:210]: I don't think we need to mention here that this is a proposed feature.

```suggestion
    via the `range` parameter (this requires support for the
```

File: runtime/doc/news.txt [hunk 4]
```diff
@@ -206,6 +206,9 @@ The following new APIs and features were added.
     the original LSP `Location` or `LocationLink`.
   • Added support for connecting to servers using named pipes (Windows) or
     unix domain sockets (Unix) via |vim.lsp.rpc.domain_socket_connect()|.
+  • |vim.lsp.buf.format()| now supports formatting passing a list of ranges
```
Comment (MariaSolOs) [inline runtime/doc/news.txt:209]:
```suggestion
  • |vim.lsp.buf.format()| now supports passing a list of ranges
```

File: runtime/lua/vim/lsp.lua [hunk 5]
```diff
@@ -1089,6 +1090,10 @@ function lsp.start_client(config)
         return true
       end
       if vim.tbl_get(client.server_capabilities, unpack(required_capability)) then
+        if method == ms.textDocument_rangesFormatting then
+          local range_formatting_provider = client.server_capabilities.documentRangeFormattingProvider
+          return type(range_formatting_provider) == 'table' and range_formatting_provider.rangesSupport == true
```
Comment (MariaSolOs) [inline runtime/lua/vim/lsp.lua:1095]:
```suggestion
          return type(range_formatting_provider) == 'table' and range_formatting_provider.rangesSupport
```

File: runtime/lua/vim/lsp/buf.lua [hunk 6]
```diff
@@ -192,17 +192,21 @@ end
...
   local mode = api.nvim_get_mode().mode
   local range = options.range
-  if not range and mode == 'v' or mode == 'V' then
+  local ranges = (range and #range ~= 0 and type(range[1]) == 'table') and range
+  if not range and not ranges and mode == 'v' or mode == 'V' then
     range = range_from_selection(bufnr, mode)
   end
-  local method = range and ms.textDocument_rangeFormatting or ms.textDocument_formatting
+  local method = (ranges and ms.textDocument_rangesFormatting) or
+      (range and ms.textDocument_rangeFormatting or ms.textDocument_formatting)
```
Comment (MariaSolOs) [inline runtime/lua/vim/lsp/buf.lua:209]: I'm not entirely sure, but if `range` is `nil` here then `method` won't be assigned. Could you please check this?

**Comment by MariaSolOs** [comment]
We would also need to add [the client capability](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.18/specification/#documentRangeFormattingClientCapabilities) here:
https://github.com/neovim/neovim/blob/2e982f1aad9f1a03562b7a451d642f76b04c37cb/runtime/lua/vim/lsp/protocol.lua#L742

File: runtime/lua/vim/lsp/buf.lua [hunk 7]
```diff
@@ -192,17 +192,21 @@ end
 ---     - range (table|nil) Range to format.
 ---         Table must contain `start` and `end` keys with {row,col} tuples using
 ---         (1,0) indexing.
+---         Can also be a list of tables that contain `start` and `end` keys as described above,
+---         in which case textDocument/rangesFormatting support is required.
```
Comment (MariaSolOs) [inline runtime/lua/vim/lsp/buf.lua:196]: What kind of error is presented to the user if the language client doesn't support `ranges`? Would it make sense to check for the capability here for a better error message?

Comment (tom-anders) [reply runtime/lua/vim/lsp/buf.lua:196]: The call to get_client() already filters out any servers that do not support the LSP method, so you'd get `[LSP] Format request failed, no matching language servers.` if none of the servers support `textDocument/rangesFormatting` (the same thing happens for `textDocument/rangeFormatting` alrady

File: runtime/lua/vim/lsp/buf.lua (hunk 6 above)
Comment (tom-anders) [reply runtime/lua/vim/lsp/buf.lua]: `:lua print((nil and "a") or (nil and "b" or "c"))` prints `c`, so I think this should work

**Review by tom-anders — COMMENTED** [review]
Thanks for the review!

File: runtime/lua/vim/lsp/buf.lua (hunk 6 above)
Comment (MariaSolOs) [reply runtime/lua/vim/lsp/buf.lua]: Thanks for checking!

File: runtime/lua/vim/lsp/buf.lua (hunk 7 above)
Comment (MariaSolOs) [reply runtime/lua/vim/lsp/buf.lua:196]: Sounds good :)

File: runtime/doc/lsp.txt (hunk 1 above)
Comment (MariaSolOs) [reply runtime/doc/lsp.txt:1261]: Sorry if I'm missing something, but this was marked as resolved yet I'm not seeing the change...?

It's okay if you decided to ignore the suggestion, but at least tell me why :)

File: runtime/doc/lsp.txt (hunk 2 above)
Comment (MariaSolOs) [reply runtime/doc/lsp.txt:1262]: Resolved? (See my comment above)

File: runtime/lua/vim/lsp.lua [hunk 8]
```diff
@@ -1089,6 +1090,12 @@ function lsp.start_client(config)
         return true
       end
       if vim.tbl_get(client.server_capabilities, unpack(required_capability)) then
+        if method == ms.textDocument_rangesFormatting then
+          local range_formatting_provider =
+            client.server_capabilities.documentRangeFormattingProvider
+          return type(range_formatting_provider) == 'table'
+            and range_formatting_provider.rangesSupport
```
Comment (MariaSolOs) [inline runtime/lua/vim/lsp.lua:1097]: I don't think this is the right pattern to deal with nested capabilities. Instead you should use a table as the value here:
https://github.com/neovim/neovim/blob/bebaf54d68ed3ebbe35003db2eeeedc59d7b56f3/runtime/lua/vim/lsp.lua#L60

Here's an example for `prepareRename`.
https://github.com/neovim/neovim/blob/cd3399f6bd99494c5805ee2560e515ec6c782e59/runtime/lua/vim/lsp.lua#L52

File: runtime/lua/vim/lsp/buf.lua [hunk 9]
```diff
@@ -192,17 +192,21 @@ end
...
 ---         Defaults to current selection in visual mode
 ---         Defaults to `nil` in other modes, formatting the full buffer
 function M.format(options)
   options = options or {}
   local bufnr = options.bufnr or api.nvim_get_current_buf()
   local mode = api.nvim_get_mode().mode
   local range = options.range
-  if not range and mode == 'v' or mode == 'V' then
+  local ranges = (range and #range ~= 0 and type(range[1]) == 'table') and range
```
Comment (MariaSolOs) [inline runtime/lua/vim/lsp/buf.lua:204]: Maybe it's just me but I find this code that deals with both scenarios of dealing with range vs ranges a bit hard to follow. I would suggest either adding some comments that explain what's going on, or clearly separating both cases with a single if-statement.

File: runtime/doc/lsp.txt (hunk 1 above)
Comment (tom-anders) [reply runtime/doc/lsp.txt:1261]: Ahrg sorry, I changed it locally in lsp.txt, but then running `make doc` overwrote it again - I changed in the right place this time (buf.lua)

File: runtime/doc/lsp.txt (hunk 2 above)
Comment (tom-anders) [reply runtime/doc/lsp.txt:1262]: same here

File: runtime/lua/vim/lsp/buf.lua (hunk 9 above)
Comment (tom-anders) [reply runtime/lua/vim/lsp/buf.lua:204]: Good point, having both `range` and `ranges` as a variable is confusing - I added some comments, but also replaced the `ranges` variable with a boolean `passed_multiple_ranges`, I think readability is much higher now

File: runtime/lua/vim/lsp.lua (hunk 8 above)
Comment (tom-anders) [reply runtime/lua/vim/lsp.lua:1097]: Ohh right, I for some reason thought `vim.tbl_get` would take adjacent keys instead of nested ones.

File: runtime/lua/vim/lsp/buf.lua [hunk 10]
```diff
@@ -192,17 +192,25 @@ end
...
   if not range and mode == 'v' or mode == 'V' then
     range = range_from_selection(bufnr, mode)
   end
-  local method = range and ms.textDocument_rangeFormatting or ms.textDocument_formatting
+  -- List of ranges => textDocument/rangesFormatting
+  -- Single range   => textDocument/rangeFormatting
+  -- No range       => textDocument/formatting
+  local passed_multiple_ranges = (range and #range ~= 0 and type(range[1]) == 'table')
+  local method = (passed_multiple_ranges and ms.textDocument_rangesFormatting)
+    or (range and ms.textDocument_rangeFormatting or ms.textDocument_formatting)
```
Comment (wookayin) [inline runtime/lua/vim/lsp/buf.lua:213]: I find this too complex to read. Let's rewrite using plain if-else?

File: runtime/lua/vim/lsp/buf.lua [hunk 11]
```diff
@@ -192,17 +192,25 @@ end
 ---     - range (table|nil) Range to format.
 ---         Table must contain `start` and `end` keys with {row,col} tuples using
 ---         (1,0) indexing.
+---         Can also be a list of tables that contain `start` and `end` keys as described above,
```
Comment (wookayin) [inline runtime/lua/vim/lsp/buf.lua:195]: It'd be better to be specific. Although the previous doc (written #19998) is also ambiguous, but it'd be better to fix in this opportunity. For normal range formatting request, we can mention `lsp.Range`; for the ranges formatting request, it would be `lsp.Range[]`.

File: runtime/lua/vim/lsp/buf.lua (hunk 10 above)
Comment (wookayin) [reply runtime/lua/vim/lsp/buf.lua:213]: Also can we have capability check? (either explicitly or implicitly) Though not mandatory, it'd be even better if this can be checked by tests.

File: runtime/lua/vim/lsp/buf.lua [hunk 12]
```diff
@@ -192,17 +192,25 @@ end
...
+---         in which case `textDocument/rangesFormatting` support is required.
 ---         Defaults to current selection in visual mode
----         Defaults to `nil` in other modes, formatting the full buffer
+---         and to `nil` in other modes, formatting the full buffer
 function M.format(options)
   options = options or {}
   local bufnr = options.bufnr or api.nvim_get_current_buf()
   local mode = api.nvim_get_mode().mode
   local range = options.range
```
Comment (wookayin) [inline runtime/lua/vim/lsp/buf.lua:203]:
```suggestion
  local range = options.range ---@type nil|lsp.Range|lsp.Range[]
```

File: runtime/lua/vim/lsp/buf.lua (hunk 11 above)
Comment (wookayin) [reply runtime/lua/vim/lsp/buf.lua:195]: Or we could name the parameter as `ranges` instead of `range` if we'd want to make it closer to the type defined in the LSP protocol (see `DocumentRangesFormattingParams`), with more strict type checking, but I don't have a string opinion here in the API design. Any thoughts, @MariaSolOs @mfussenegger?

File: runtime/lua/vim/lsp/buf.lua [hunk 13]
```diff
@@ -219,10 +227,14 @@ function M.format(options)
...
-    if range then
-      local range_params =
-        util.make_given_range_params(range.start, range['end'], bufnr, client.offset_encoding)
-      params.range = range_params.range
+    local to_lsp_range = function(r)
+      return util.make_given_range_params(r.start, r['end'], bufnr, client.offset_encoding).range
+    end
+
+    if passed_multiple_ranges then
+      params.ranges = vim.tbl_map(to_lsp_range, range)
```
Comment (wookayin) [inline runtime/lua/vim/lsp/buf.lua:235]: Please add typing (`lsp.DocumentRangeFormattingParams` or `lsp.DocumentRangesFormattingParams`) to this local function.

**Review by wookayin — COMMENTED** [review]
This would be a great enhancement to the built-in LSP. Please also see `:help lsp-method` `:help vim.lsp.util.make_range_params()`.

File: runtime/lua/vim/lsp/buf.lua (hunk 11 above)
Comment (MariaSolOs) [reply runtime/lua/vim/lsp/buf.lua:195]: I agree that using `lsp.Range` here is nicer/safer. I don't have any preference regarding naming.

**Comment by MariaSolOs** [comment]
Pinging @stevearc in case this impacts `conform` in any way, or there's something you could take advantage of :)

File: runtime/lua/vim/lsp/buf.lua (hunk 10 above)
Comment (tom-anders) [reply runtime/lua/vim/lsp/buf.lua:213]: > I find this too complex to read. Let's rewrite using plain if-else?
Done!

> Also can we have capability check? (either explicitly or implicitly) Though not mandatory, it'd be even better if this can be checked by tests.

Our call to `vim.lsp.get_clients` already filters by method, so there's already a (somewhat hidden) capability check there,

File: runtime/lua/vim/lsp/buf.lua (hunk 11 above)
Comment (tom-anders) [reply runtime/lua/vim/lsp/buf.lua:195]: Hmm the thing is that this is **not** an LSP range

The function expects a range of the form  `{ start = {row, col}, end = {row, col} }` with `row` being 1-indexed, but `col` being 0-indexed (for some reason...?)

Meanwhile, an [LSP range](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#range) is `{ start = { line = .., character = ..}, end = { line =..., character = ...}}`, with both `line` and `character` using 0-indexing.

Only the call to `utils.make_given_range_params` turns our "custom" range into a LSP range.

I'd also find it more intuitive if this would take a LSP range instead, but I think there are two problems:

1) API break (obviously)
2) we don't have a (convenient) API yet to create LSP range objects, I think...? The closest match is `make_given_range_params`, which returns (as the name says) a `RangeParams` LSP object, not a `Range`. If we're fine with the API break, I think we should add a `vim.lsp.util.make_give_range`

Comment (wookayin) [reply runtime/lua/vim/lsp/buf.lua:195]: I see. My bad, sorry for confusing. (This is the reason why it has to be more clearly documented!) You're right -- the parameter isn't already a LSP range, but *(1,0)-indexed* API (row, col) representation; then we should not change and break the API.

Comment (MariaSolOs) [reply runtime/lua/vim/lsp/buf.lua:195]: Yeah, https://github.com/neovim/neovim/issues/25509 is a tracking issue for this problem.

Comment (tom-anders) [reply runtime/lua/vim/lsp/buf.lua:195]: Right, I'll add a comment along the lines of (this is **not** an LSP range) to the docs, if everyone's fine with that

Comment (MariaSolOs) [reply runtime/lua/vim/lsp/buf.lua:195]: I don't think such a remark is needed. The longer the docs are, the less chances people will actually read them. I think that the current description is good enough.

**Comment by tom-anders** [comment]
(ran `make doc` again to make CI happy)

**Comment by tom-anders** [comment]
(Rebased to fix news.txt conflict once again)

**Review by justinmk — APPROVED** [review]

**Comment by tom-anders** [comment]
Rebased to fix conflicts in news.txt - Anything preventing this from getting merged?

File: runtime/doc/lsp.txt [hunk 14]
```diff
@@ -1371,11 +1372,14 @@ format({opts})                                          *vim.lsp.buf.format()*
                   (client.id) matching this field.
                 • {name}? (`string`) Restrict formatting to the client with
                   name (client.name) matching this field.
-                • {range}? (`{start:integer[],end:integer[]}`, default:
-                  current selection in visual mode, `nil` in other modes,
-                  formatting the full buffer) Range to format. Table must
-                  contain `start` and `end` keys with {row,col} tuples using
-                  (1,0) indexing.
+                • {range}?
+                  (`{start:integer[],end:integer[]}|{start:integer[],end:integer[]}[]`,
```
Comment (MariaSolOs) [inline runtime/doc/lsp.txt:1376]: Would it make more sense to change these to a tuple like `[integer, integer]`?

Comment (tom-anders) [reply runtime/doc/lsp.txt:1376]: I'm not 100% sure what you mean, could you give an example of what that would look Iike?

Comment (MariaSolOs) [reply runtime/doc/lsp.txt:1376]: I mean making the type of the positions a tuple instead of `integer[]`, since we're expecting their length to be 2.

Comment (tom-anders) [reply runtime/doc/lsp.txt]: ah got it, thanks for clarifying

**Comment by tom-anders** [comment]
Last build seemed to fail due to github issues, so I did another force push (without any changes) to trigger CI again - Looks better now, not sure what's up with the BSD build...?

**Comment by tom-anders** [comment]
> Last build seemed to fail due to github issues, so I did another force push (without any changes) to trigger CI again - Looks better now, not sure what's up with the BSD build...?

BSD build is fixed, yay!
